### PR TITLE
Mark applicable structs as readonly and use in-modifier

### LINKED ISSUE #13915: Mark non-mutable structs readonly
_Tracking issue_

### PR DESCRIPTION
**Pull request by xtqqczze**
Work toward #13915.

Changes touch public APIs but do not cause any overload resolution changes, so in theory are non-breaking:
* https://blog.paranoidcoding.com/2019/03/27/readonly-struct-breaking-change.html
* https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-parameter-modifier#:~:text=The%20in%2C%20ref%2C%20and%20out%20keywords%20are%20not,The%20following%20code%2C%20for%20example%2C%20will%20not%20compile%3A

See also: https://github.com/dotnet/coreclr/pull/14789

### CODE REVIEW

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventDescriptor.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 namespace System.Diagnostics.Eventing
 {
     [StructLayout(LayoutKind.Explicit, Size = 16)]
-    public struct EventDescriptor
+    public readonly struct EventDescriptor
```
Comment (rjmholt) [inline src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventDescriptor.cs:11]: I'm not sure about changing public structs to readonly -- existing code that relies on the previous form may be affected

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs [hunk 2]
```diff
@@ -488,9 +488,9 @@ public bool WriteMessageEvent(string eventMessage)
         /// </param>
         /// <param name="eventPayload">
         /// </param>
-        public bool WriteEvent(ref EventDescriptor eventDescriptor, params object[] eventPayload)
+        public bool WriteEvent(in EventDescriptor eventDescriptor, params object[] eventPayload)
```
Comment (rjmholt) [inline src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs:491]: Changing the signature on a public method like this is a breaking change

File: src/System.Management.Automation/engine/MshCmdlet.cs [hunk 3]
```diff
@@ -68,9 +68,9 @@ public interface IDynamicParameters
     /// Type used to define a parameter on a cmdlet script of function that
     /// can only be used as a switch.
     /// </summary>
-    public struct SwitchParameter
+    public readonly struct SwitchParameter
```
[the remark this answers is not in the document]
Comment (xtqqczze) [reply src/System.Management.Automation/engine/MshCmdlet.cs:71]: But `SwitchParameter` is immutable by design, and adding the readonly-modifier doesn't cause any overload resolution changes [1], so I think this is not a breaking change.

[1] https://blog.paranoidcoding.com/2019/03/27/readonly-struct-breaking-change.html

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventDescriptor.cs (hunk 1 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventDescriptor.cs:11]: This struct is immutable by design.

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs (hunk 2 above)
Comment (xtqqczze) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs:491]: The compiler doesn't permit overloading on ref/out/in differences, so this change is non-breaking.

File: src/System.Management.Automation/engine/MshCmdlet.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rjmholt) [reply src/System.Management.Automation/engine/MshCmdlet.cs:71]: Ok I'm fairly convinced by this

**Comment by rjmholt** [comment]
@daxian-dbw can you review please?

File: src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs (hunk 2 above)
Comment (rjmholt) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs:491]: Technically the issue isn't with the compiler, but with the runtime. Modules compiled against the old API surface must continue to work after any changes are made to the API.

But I tested it out with a toy project (an exe and a lib, where I built an `in` version of the lib and then changed it so the exe is compiled against a `ref` version) and it seems to work.

Comment (xtqqczze) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs:491]: [The in, ref, and out keywords are not considered part of the method signature for the purpose of overload resolution. Therefore, methods cannot be overloaded if the only difference is that one method takes a ref or in argument and the other takes an out argument. ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/in-parameter-modifier#:~:text=The%20in%2C%20ref%2C%20and%20out%20keywords%20are%20not,The%20following%20code%2C%20for%20example%2C%20will%20not%20compile%3A)

Comment (rjmholt) [reply src/Microsoft.PowerShell.CoreCLR.Eventing/DotNetCode/Eventing/EventProvider.cs:491]: Yeah that's still C# compiler rules, where the compiler statically decides which overload to resolve for a call. My concern was that the resolved method call in the IL might not match what the library was originally compiled against when its loaded at runtime (so even though the original program isn't broken it still requires recompilation, which is a dealbreaker for modules). It makes sense that that's not the case though.

**Review by rjmholt — APPROVED** [review]

**Pull request merged**